Framework/PythonInterface/plugins/algorithms/HB2AReduce.py

```diff
@@ -0,0 +1,274 @@
+        self.declareProperty('BinData', True, "Data will be binned using BinWidth. If False then all data will be unbinned")
+        self.setPropertySettings("BinData", condition)
+        positiveFloat = FloatBoundedValidator(lower=0., exclusive=True)
+        self.declareProperty('BinWidth', 0.05, positiveFloat, "Bin size of the output workspace")
+        self.setPropertySettings("BinWidth", condition)
+        self.declareProperty('Scale', 1.0, positiveFloat, "The output will be scaled by this value")
+        self.declareProperty(WorkspaceProperty("OutputWorkspace", "",
+                                               optional=PropertyMode.Mandatory,
+                                               direction=Direction.Output),
+                             "Output Workspace")
+
+    def validateInputs(self):
+        issues = dict()
+
+        if not self.getProperty("Filename").value:
+            if ((self.getProperty("IPTS").value == Property.EMPTY_INT) or
+                (self.getProperty("Exp").value == Property.EMPTY_INT) or
+               len(self.getProperty("ScanNumbers").value) is 0):
+                issues["Filename"] = 'Must specify either Filename or IPTS AND Exp AND ScanNumbers'
+
+        return issues
+
+    def PyExec(self):
+        scale = self.getProperty("Scale").value
+        filenames = self.getProperty("Filename").value
+
+        if not filenames:
+            ipts = self.getProperty("IPTS").value
+            exp = self.getProperty("Exp").value
+            filenames = ['/HFIR/HB2A/IPTS-{0}/exp{1}/Datafiles/HB2A_exp{1:04}_scan{2:04}.dat'.format(ipts, exp, scan)
+                         for scan in self.getProperty("ScanNumbers").value]
+
+        # Read in data array and append all files
+        for n, filename in enumerate(filenames):
+            # Read in all lines once
+            with open(filename) as f:
+                lines = f.readlines()
+
+            if n==0:
+                # Read in metadata from first file only file
+                metadata = dict([np.char.strip(re.split('#(.*?)=(.*)', line, flags=re.U)[1:3])
+                                 for line in lines if re.match('^#.*=', line)])
+                # Get indir and exp from first file
+                indir, data_filename = os.path.split(filename)
+                _, exp, _ = data_filename.replace(".dat", "").split('_')
+
+            # Find size of header, the size changes
+            header = np.argmax([bool(re.match('(?!^#)', line)) for line in lines])-1
+            if header < 0:
+                raise RuntimeError("{} has no data in it".format(filename))
+            names = lines[header].split()[1:]
+
+            try:
+                d = np.loadtxt(lines[header:], ndmin=1, dtype={'names': names, 'formats':[float]*len(names)})
+            except (ValueError, IndexError):
+                raise RuntimeError("Count not read {}, file likely malformed".format(filename))
+
+            try:
```

Comment:
This would be better with `data = None` outside the loop and a check for it here. Then the standard workflow wouldn't involve handling an exception.